sklearn/feature_selection/univariate_selection.py

```diff
@@ -227,6 +230,188 @@ def chi2(X, y):
     return _chisquare(observed, expected)
 
 
+def get_entropy(prob):
+    t = np.log2(prob)
+    t[~np.isfinite(t)] = 0
+    return np.multiply(-prob, t)
+
+
+def _info_gain(X, y, aggregate, ratio=False):
+
+    def _a_log_a_div_b(a, b):
+        with np.errstate(invalid='ignore', divide='ignore'):
+            t = np.log2(a / b)
+        t[~np.isfinite(t)] = 0
+        return np.multiply(a, t)
+
+    f_count, c_count, fc_count, total = _get_fc_counts(X, y)
+
+    f_prob = f_count / f_count.sum()
+    c_prob = c_count / c_count.sum()
+    fc_prob = fc_count / total
+
+    c_f = _a_log_a_div_b(fc_prob, c_prob * f_prob)
+    nc_nf = _a_log_a_div_b(1 - f_prob - c_prob + fc_prob,
+                           (1 - c_prob) * (1 - f_prob))
+    c_nf = _a_log_a_div_b((c_count - fc_count) / total, c_prob * (1 - f_prob))
+    nc_f = _a_log_a_div_b((f_count - fc_count) / total, (1 - c_prob) * f_prob)
+
+    scores = c_f + nc_nf + c_nf + nc_f
+
+    if ratio:
+        # normalize IG scores to obtain GR
+        with np.errstate(invalid='ignore', divide='ignore'):
+            scores = scores / (get_entropy(c_prob) + get_entropy(1 - c_prob))
+
+    # the feature score is averaged over classes
+    if aggregate == "mean":
+        scores = scores.mean(axis=0)
+    elif aggregate == "sum":
+        scores = scores.sum(axis=0)
+    else:
+        scores = scores.max(axis=0)
+
+    return np.asarray(scores).reshape(-1), []
+
+
+def _get_fc_counts(X, y):
+    """Count feature, class, joint and total frequencies
+
+    Returns
+    -------
+    f_count : array, shape = (n_features,)
+    c_count : array, shape = (n_classes,)
+    fc_count : array, shape = (n_features, n_classes)
+    total: int
+    """
+    X = check_array(X, accept_sparse=['csr', 'coo'])
```

Comment:
Why 'coo'? I think 'csc' might be most appropriate: I think X will be converted into CSC when `safe_sparse_dot(Y.T, X)` is performed below (assuming Y is CSR, as output by `LabelBinarizer`).